### PR TITLE
Add errors to peer store

### PR DESCRIPTION
**Pull request by n8maninger**

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -42,17 +44,28 @@ type PeerInfo struct {
...
+	UpdatePeerInfo(peer string, fn func(*PeerInfo)) error
 	// Ban temporarily bans one or more IPs. The addr should either be a single
 	// IP with port (e.g. 1.2.3.4:5678) or a CIDR subnet (e.g. 1.2.3.4/16).
-	Ban(addr string, duration time.Duration, reason string)
-	Banned(peer string) bool
+	Ban(addr string, duration time.Duration, reason string) error
+
+	// Banned returns nil if the peer is not banned. Otherwise, ErrPeerBanned
+	// is returned.
+	Banned(peer string) error
```
Comment (lukechampine) [inline syncer/syncer.go:61]: sentinel values are tempting, but I prefer `(bool, error)` for this sort of thing. Compare:
```go
if banned, err := ps.Banned(peer); err != nil {
    // handle err
} else if banned {
    // handle banned
}
```
vs.
```go
if err := ps.Banned(peer); err != nil && err != ErrPeerBanned {
    // handle err
} else if err == ErrPeerBanned {
    // handle banned
}

// or: 

if err := ps.Banned(peer); err == ErrPeerBanned {
    // handle banned
} else if err != nil {
    // handle err
}

// or:

if ps.Banned(peer) == ErrPeerBanned {
    // handle banned
} else if ps.Banned(peer) != nil {
    // handle err
}
```
all of these are dissatisfactory imo.

File: syncer/syncer.go [hunk 2]
```diff
@@ -42,17 +44,27 @@ type PeerInfo struct {
...
+	// UpdatePeerInfo updates the metadata for the specified peer. If the peer
+	// is not found, the error should be ErrPeerNotFound.
+	UpdatePeerInfo(addr string, fn func(*PeerInfo)) error
 	// Ban temporarily bans one or more IPs. The addr should either be a single
 	// IP with port (e.g. 1.2.3.4:5678) or a CIDR subnet (e.g. 1.2.3.4/16).
-	Ban(addr string, duration time.Duration, reason string)
-	Banned(peer string) bool
+	Ban(addr string, duration time.Duration, reason string) error
+
+	// Banned returns true,nil if the peer is banned.
```
Comment (peterjan) [inline syncer/syncer.go:59]:
```suggestion
	// Banned returns true, nil if the peer is banned.
```

File: syncer/syncer.go [hunk 3]
```diff
@@ -42,17 +44,27 @@ type PeerInfo struct {
...
+	Banned(addr string) (bool, error)
 }
 
+var (
+	// ErrPeerBanned is returned when a peer is banned.
+	ErrPeerBanned = errors.New("peer is banned")
```
Comment (peterjan) [inline syncer/syncer.go:65]: is this now unused?

File: syncer/syncer.go [hunk 4]
```diff
@@ -194,14 +206,16 @@ func (s *Syncer) resync(p *Peer, reason string) {
 	}
 }
 
-func (s *Syncer) ban(p *Peer, err error) {
+func (s *Syncer) ban(p *Peer, err error) error {
 	s.log.Debug("banning peer", zap.Stringer("peer", p), zap.Error(err))
 	p.setErr(errors.New("banned"))
-	s.pm.Ban(p.ConnAddr, 24*time.Hour, err.Error())
+	if err := s.pm.Ban(p.ConnAddr, 24*time.Hour, err.Error()); err != nil {
```
Comment (peterjan) [inline syncer/syncer.go:212]: `nit` could/should be a constant

File: syncer/syncer.go [hunk 5]
```diff
@@ -218,17 +232,25 @@ func (s *Syncer) ban(p *Peer, err error) {
 			s.strikes[subnet]++
 		}
 		s.mu.Unlock()
-		if ban {
-			s.pm.Ban(subnet, 24*time.Hour, "too many strikes")
+		if !ban {
+			continue
+		} else if err := s.pm.Ban(subnet, 24*time.Hour, "too many strikes"); err != nil {
```
Comment (peterjan) [inline syncer/syncer.go:237]: can reuse the constant here already, maybe reuse `ErrPeerBanned` here and update the error message, I feel this error should be exported

**Review by ChrisSchinnerl — APPROVED** [review]

**Review by ChrisSchinnerl — APPROVED** [review]

**Pull request merged**